ts/defs.bzl

```diff
@@ -68,7 +72,169 @@ def ts_project(
         **kwargs):
     """Compiles one TypeScript project using `tsc --project`.
 
-    For the list of args, see the ts_project rule.
+    This is a drop-in replacement for the `tsc` rule automatically generated for the "typescript"
+    package, typically loaded from `@npm//typescript:package_json.bzl`.
+    Unlike bare `tsc`, this rule understands the Bazel interop mechanism (Providers)
+    so that this rule works with others that produce or consume TypeScript typings (`.d.ts` files).
+
+    Some TypeScript options affect which files are emitted, and Bazel needs to predict these ahead-of-time.
+    As a result, several options from the tsconfig file must be mirrored as attributes to ts_project.
+    A validator action is run to help ensure that these are correctly mirrored.
+    See https://www.typescriptlang.org/tsconfig for a listing of the TypeScript options.
+
+    Any code that works with `tsc` should work with `ts_project` with a few caveats:
+
+    - ts_project` always produces some output files, or else Bazel would never run it.
+      Therefore you shouldn't use it with TypeScript's `noEmit` option.
+      If you only want to test that the code typechecks, instead use
+      ```
+      load("@npm//typescript:package_json.bzl", "bin")
+      bin.tsc_test( ... )
+      ```
+    - Bazel requires that the `outDir` (and `declarationDir`) be set beneath
+      `bazel-out/[target architecture]/bin/path/to/package`
+      so we override whatever settings appear in your tsconfig.
```

Comment:
Perhaps this could be a bit more prescriptive? Something like...

Do not set `outDir` and `declarationDir`. `ts_project` overrides these settings to point beneath `bazel-out/[target architecture]/bin/path/to/package`.